tsl/src/nodes/vector_agg/exec.c

```diff
@@ -0,0 +1,204 @@
+{
+	CustomScan *cscan = castNode(CustomScan, node->ss.ps.plan);
+	node->custom_ps =
+		lappend(node->custom_ps, ExecInitNode(linitial(cscan->custom_plans), estate, eflags));
+}
+
+static void
+vector_agg_end(CustomScanState *node)
+{
+	ExecEndNode(linitial(node->custom_ps));
+}
+
+static void
+vector_agg_rescan(CustomScanState *node)
+{
+	if (node->ss.ps.chgParam != NULL)
+		UpdateChangedParamSet(linitial(node->custom_ps), node->ss.ps.chgParam);
+
+	ExecReScan(linitial(node->custom_ps));
+}
+
+static TupleTableSlot *
+vector_agg_exec(CustomScanState *vector_agg_state)
+{
+	DecompressChunkState *decompress_state =
+		(DecompressChunkState *) linitial(vector_agg_state->custom_ps);
+
+	/*
+	 * The aggregated targetlist with Aggrefs is in the custom scan targetlist
+	 * of the custom scan node that is performing the vectorized aggregation.
+	 * We do this to avoid projections at this node, because the postgres
+	 * projection functions complain when they see an Aggref in a custom
+	 * node output targetlist.
+	 * The output targetlist, in turn, consists of just the INDEX_VAR references
+	 * into the custom_scan_tlist.
+	 */
+	List *aggregated_tlist = castNode(CustomScan, vector_agg_state->ss.ps.plan)->custom_scan_tlist;
+	Assert(list_length(aggregated_tlist) == 1);
+
+	/* Checked by planner */
+	Assert(ts_guc_enable_vectorized_aggregation);
+	Assert(ts_guc_enable_bulk_decompression);
+
+	/* Determine which kind of vectorized aggregation we should perform */
+	TargetEntry *tlentry = (TargetEntry *) linitial(aggregated_tlist);
+	Assert(IsA(tlentry->expr, Aggref));
+	Aggref *aggref = castNode(Aggref, tlentry->expr);
+
+	/* Partial result is a int8 */
+	Assert(aggref->aggtranstype == INT8OID);
+
+	Assert(list_length(aggref->args) == 1);
+
+	/* The aggregate should be a partial aggregate */
+	Assert(aggref->aggsplit == AGGSPLIT_INITIAL_SERIAL);
+
+	Var *var = castNode(Var, castNode(TargetEntry, linitial(aggref->args))->expr);
+
+	DecompressContext *dcontext = &decompress_state->decompress_context;
+
+	CompressionColumnDescription *value_column_description = NULL;
+	for (int i = 0; i < dcontext->num_total_columns; i++)
+	{
+		CompressionColumnDescription *current_column = &dcontext->template_columns[i];
+		if (current_column->output_attno == var->varattno)
+		{
+			value_column_description = current_column;
+			break;
+		}
+	}
+	Ensure(value_column_description != NULL, "aggregated compressed column not found");
+
+	Assert(value_column_description->type == COMPRESSED_COLUMN ||
+		   value_column_description->type == SEGMENTBY_COLUMN);
+
+	BatchQueue *batch_queue = decompress_state->batch_queue;
+	DecompressBatchState *batch_state = batch_array_get_at(&batch_queue->batch_array, 0);
+
+	/* Get a reference the the output TupleTableSlot */
+	TupleTableSlot *aggregated_slot = vector_agg_state->ss.ps.ps_ResultTupleSlot;
+	Assert(aggregated_slot->tts_tupleDescriptor->natts == 1);
+
+	Assert(aggref->aggfnoid == F_SUM_INT4);
```

Comment:
This function now contains the generic part of the node and the sum function. Is there a specific reason why it is merged in this large function? 

So far, the sum function implementation was part of `perform_vectorized_sum_int4` and we had some generic code to choose the actual implementation. This made it easy to enhance the node with further custom implementations.

Reply:
Right, removed the checks specific for int4sum.